### PR TITLE
sandbox/cgroup: support killing running snap apps in snapd directly

### PR DESCRIPTION
**Pull request by ZeyadYasser**
A problem was identified with the app terminating logic that depends on the user session agent (https://github.com/canonical/snapd/pull/14160) which would make it impossible to use on Ubuntu Core and apps started by root since user session agent won't be running.

This PR does the following:
- Implement a new terminate algorithm where snapd identifies running processes cgroups and kills them directly
- Update backend to use new terminate algorithm instead of userd based implementation
- Revert the userd implementation
- Add a spread test that runs everywhere (cgroup v1, cgroup v2, classic, core) as root.

This PR should unblock https://github.com/canonical/snapd/pull/14192 once merged.

### CODE REVIEW

File: sandbox/cgroup/freezer.go
Comment (zyga) [inline sandbox/cgroup/freezer.go:191]: Given that we are changing this, I think we should not do this anymore.

Instead we should:

- accept a context
- allow the caller to cancel the context
- poll indefinitely until the change is done or until the context is cancelled

File: sandbox/cgroup/kill.go
Comment (zyga) [inline sandbox/cgroup/kill.go:48]: Can you leave a TODO that this is racy and that instead we should open each pid, verify the cgroup it inhabits and then kill it only if the cgroup still matches. As written we have no guarantee that the PID we are racing to kill is the one from the cgroup due to the silly design in old Linux that has recently finally been fixed with pidfd.

File: sandbox/cgroup/kill.go
Comment (zyga) [inline sandbox/cgroup/kill.go:96]: Could we use errors.Join now?

**Review by zyga — COMMENTED** [review]
Quick few comments. Have a look.

File: sandbox/cgroup/freezer.go
Comment (zyga) [reply sandbox/cgroup/freezer.go:191]: This way the caller can control the timeout and it's not so fragile anymore.

File: sandbox/cgroup/kill.go
Comment (ZeyadYasser) [reply sandbox/cgroup/kill.go:48]: I did some digging into this, this is what I found:
- For v1:
  - The pattern of `freeze -> kill -> thaw` is race free because the v1 freezer puts processes in an uninterruptable sleep, so they cannot die even if SIGKILL was sent externally.
- For v2:
  - It seems that the v2 freezer was only available since [Linux 5.2](https://kernelnewbies.org/Linux_5.2#Add_freezer_controller_for_cgroups_v2) so freezing is a no-op before 5.2
  - Also, freezing doesn't put processes in an uninterruptable sleep, so they can be killed externally and have their pid reused
  - The good news is that `pidfd_send_signal` is supported in [Linux 5.1](https://kernelnewbies.org/Linux_5.1#Safe_signal_delivery_in_presence_of_PID_reuse) so starting from 5.2 we can reliably use `freezing + pidfds` but I am not sure it is worth the extra complexity as `cgroup.kill` for v2 was introduced in [Linux 5.14](https://kernelnewbies.org/Linux_5.14#cgroup_kill_interface_to_kill_all_processes) and solves all our problems on relatively newer kernels (after Aug 2021).
  - We now have this window between [Linux 4.5](https://kernelnewbies.org/Linux_4.5#cgroup_unified_hierarchy_is_considered_stable) where v2 was introduced as stable and  [Linux 5.2](https://kernelnewbies.org/Linux_5.2#Add_freezer_controller_for_cgroups_v2) where v2 freezer was introduced where it seems it is unavoidable to have a race if v2 was used even if we add the pidfds complexity.

A note on distributions that enable v2 by default according to [runc](https://github.com/opencontainers/runc/blob/main/docs/cgroup-v2.md#cgroup-v2):
- Fedora (since 31): **31 is EOL, The currently supported 39 uses Linux 6.5, so we are good there**
- Arch Linux (since April 2021): **Rolling release, most likely using latest kernel**
- openSUSE Tumbleweed (since c. 2021): **Rolling release, currently at [6.10.4](https://software.opensuse.org/package/kernel-default), openSUSE Leap 15.5 uses [5.14](https://en.opensuse.org/Features_15.5#Linux_kernel) while openSUSE Leap 15.6 uses [6.4](https://news.opensuse.org/2024/03/07/leap-reaches-beta-phase/)**
- Debian GNU/Linux (since 11): **Debian 11 uses [Linux 5.10](https://www.debian.org/releases/bullseye/amd64/release-notes/ch-whats-new.en.html#major-packages) so we would be missing `cgroup.kill`, Debian 12 uses [6.1](https://www.debian.org/releases/bookworm/amd64/release-notes/ch-whats-new.en.html#major-packages) though**
- Ubuntu (since 21.10): **21.10 is EOL, 22.04 uses [Linux 5.15](https://ubuntu.com/kernel/lifecycle) so we would have `cgroup.kill`**
- RHEL and RHEL-like distributions (since 9): **Centos 9 uses Linux 5.14**
- Amazon Linux: **AML2 ships with both [4.15 and (default) 5.10](https://docs.aws.amazon.com/linux/al2/ug/aml2-kernel.html) (EOL 2025-06-30) while AML2023 brings [6.1](https://docs.aws.amazon.com/linux/al2023/ug/compare-with-al2-kernel.html)**

A similar discussion can be found in [systemd](https://github.com/systemd/systemd/issues/13101).

@pedronis, @zyga What are your thoughts?

Comment (ZeyadYasser) [reply sandbox/cgroup/kill.go:48]: My conclusion is using `pidfds` + `pidfd_send_signal` + `confirm pidfd belongs to cgroup` would solve the race for v2 in the real world since most distros that default to using v2 use a 5.2+ kernel and anything 5.14+ will fallback to the simpler `cgroup.kill`.

I think now the question is, should we add this complexity?

File: sandbox/cgroup/freezer.go
Comment (ZeyadYasser) [reply sandbox/cgroup/freezer.go:191]: I would like to add an internal timeout as well, so that if the caller doesn't pass a timeout context we still exit if cgroup is stuck in FREEZING state, having an upper bound of 5-10s.

File: sandbox/cgroup/kill.go
Comment (ZeyadYasser) [reply sandbox/cgroup/kill.go:48]: Actually we could tackle the window between Linux 4.5 and 5.2 where we cannot freeze a cgroup by doing multiple iterations when SIGKILL is passed  until the cgroup is drained of all procs even if the process keeps forking similar to what systemd is doing [here](https://github.com/systemd/systemd/blob/98b86d294adcca9887d0676190e7f697f45f7f54/src/basic/cgroup-util.c#L386-L387).

But that would also mean adding to already existing complexity of pidfds.

File: sandbox/cgroup/kill.go
Comment (ZeyadYasser) [reply sandbox/cgroup/kill.go:96]: master is still at go 1.18, errors.Join was introduced in 1.20

File: sandbox/cgroup/kill.go
Comment (zyga) [reply sandbox/cgroup/kill.go:67]: I think this is a fair assesment and I would suggest that we use `cgroup.kill` and fall back to iterating over processes like we do, with a task to perhaps add pidfd support as a separate task we file in Jira and schedule sometime during this year but not for 2.65.

File: sandbox/cgroup/freezer.go
Comment (zyga) [inline sandbox/cgroup/freezer.go:63]: Perhaps mention maxFreezeTimeout instead of the actual value.

File: sandbox/cgroup/freezer.go
Comment (zyga) [inline sandbox/cgroup/freezer.go:209]: This is deprecated, use `errors.Is(err, fs.ErrNotExist)`

File: sandbox/cgroup/freezer.go
Comment (zyga) [inline sandbox/cgroup/freezer.go:199]: I think we can preserve the original error here.

```suggestion
		return fmt.Errorf("cannot freeze processes in group %q: %w", groupName, err)
```

File: sandbox/cgroup/freezer.go
Comment (zyga) [inline sandbox/cgroup/freezer.go:265]: `os.IsNotExist` is deprecated.

File: sandbox/cgroup/kill.go
Comment (zyga) [inline sandbox/cgroup/kill.go:43]: How is this selected at runtime?

Separately, I think we should default to v2 and only fallback to v1 if an expected interface file is missing.

File: sandbox/cgroup/kill.go
Comment (zyga) [inline sandbox/cgroup/kill.go:81]: I think thaw may be duplicated in one of the inner functions. If we thaw here, let's drop the thaw from the lower level functions (or vice-versa but given that we freeze here it looks better).

File: sandbox/cgroup/kill.go
Comment (zyga) [inline sandbox/cgroup/kill.go:97]: `os.IsNotExist` is deprecated.

File: sandbox/cgroup/kill.go
Comment (zyga) [inline sandbox/cgroup/kill.go:111]: Ditto

File: sandbox/cgroup/kill.go
Comment (zyga) [inline sandbox/cgroup/kill.go:94]: Do we ever send other signals? Perhaps we can just require the signal to be SIGKILL and drop the logic?

File: sandbox/cgroup/kill.go
Comment (zyga) [inline sandbox/cgroup/kill.go:105]: Should we only thaw on error?

File: sandbox/cgroup/kill.go
Comment (zyga) [reply sandbox/cgroup/kill.go:81]: Separately, should we only thaw on error?

File: tests/main/snap-remove-terminate/task.yaml
Comment (zyga) [inline tests/main/snap-remove-terminate/task.yaml:26]: How does this make sure the process is alive?

File: tests/main/snap-remove-terminate/task.yaml
Comment (zyga) [inline tests/main/snap-remove-terminate/task.yaml:23]: We have some tests that use `systemd-run` instead of hand-holding the process like that. There were countless issues with odd race conditions and stuck file descriptors with code running like that, using shell job control primitives.

File: tests/main/snap-remove-terminate/task.yaml
Comment (zyga) [inline tests/main/snap-remove-terminate/task.yaml:35]: Somewhat of a nitpick but this is just as racy. Perhaps the process should be a flock and we should ensure that the locked file is no longer locked. Together with systemd-run we could make this test PID-free.

**Review by zyga — COMMENTED** [review]
I left a number of comments inline. Let's sync on details or just ping me the moment you want to talk or have comments or code updates.

File: sandbox/cgroup/kill.go
Comment (ZeyadYasser) [reply sandbox/cgroup/kill.go:43]: It is picked in the init for the cgroup package
https://github.com/canonical/snapd/blob/df2ffc71505a3a7a0566bb0c61bf5c81c0c0177b/sandbox/cgroup/cgroup.go#L64-L72
https://github.com/canonical/snapd/pull/14354/files/089e9df70b9dea0ecd9c74f7033d57bb10283713#diff-aab845d18e1a63cdd61215ac4596638ea81e894078faaff2a75c7f6cd52641aeR77

File: sandbox/cgroup/kill.go
Comment (ZeyadYasser) [reply sandbox/cgroup/kill.go:94]: no, we currently don't. I can drop it for now and if we ever need other signals we can add support later

File: sandbox/cgroup/kill.go
Comment (ZeyadYasser) [reply sandbox/cgroup/kill.go:81]: `freezeSnapProcessesImplV1` internally only thaws if there is an error to avoid having a side effect on failure. 

For V1 we must thaw a frozen cgroup for the SIGKILL to take effect. It might be a good idea to document this kernel behavior in a comment.

File: sandbox/cgroup/kill.go
Comment (ZeyadYasser) [reply sandbox/cgroup/kill.go:105]: Nice catch, we only need to thaw on error here unlike in V1

File: tests/main/snap-remove-terminate/task.yaml
Comment (ZeyadYasser) [reply tests/main/snap-remove-terminate/task.yaml:26]: Signal 0 doesn't send a signal but does error checking that the pid exists https://linux.die.net/man/2/kill

File: tests/main/snap-remove-terminate/task.yaml
Comment (ZeyadYasser) [reply tests/main/snap-remove-terminate/task.yaml]: Thanks for the suggestion, the test is much cleaner/stable now

File: sandbox/cgroup/kill.go
Comment (zyga) [reply sandbox/cgroup/kill.go:84]: Ah, that is a very good point. Please leave a comment that makes this clear in the code.

File: sandbox/cgroup/kill.go
Comment (zyga) [reply sandbox/cgroup/kill.go]: Yeah, let's reduce this. Simple is good.

File: tests/main/snap-remove-terminate/task.yaml
Comment (ZeyadYasser) [reply tests/main/snap-remove-terminate/task.yaml]: This was just checking that the background process is up, this is now replaced with systemd-run

File: cmd/snap-update-ns/common.go
Comment (zyga) [inline cmd/snap-update-ns/common.go:83]: I wonder if we should bubble up this all the way to the top. Ideally we'd cancell the context with `signal.NotifyContext`. Perhaps leave a TODO or even use `context.TODO` to make it clear why.

File: cmd/snap-update-ns/common_test.go
Comment (zyga) [inline cmd/snap-update-ns/common_test.go:27]: This is probably automatically added by your editor. Please use `context` from the standard library.

File: overlord/snapstate/backend/link.go
Comment (zyga) [inline overlord/snapstate/backend/link.go:421]: I wonder if we should use a `context.TODO` and perhaps bubble it up somehow. It's also interesting, from task engine perspecitve, that tasks might want to allow to be cancelled this way.

File: overlord/snapstate/handlers.go
Comment (zyga) [inline overlord/snapstate/handlers.go:3317]: Is this so that we match with `defer st.Unlock()` at the entry to the function?

File: sandbox/cgroup/freezer.go
Comment (zyga) [inline sandbox/cgroup/freezer.go:91]: Given that `errors.Is` checks the error, I think it's fine to remove `err != nil` condition (here and elsewhere).

File: sandbox/cgroup/freezer.go
Comment (zyga) [inline sandbox/cgroup/freezer.go:103]: should this be a variable as well? It might be good to mock it in tests perhaps.

File: sandbox/cgroup/freezer.go
Comment (zyga) [inline sandbox/cgroup/freezer.go:207]: Ditto. Perhaps call this `freezePulseDelay`?

Comment (zyga) [inline sandbox/cgroup/freezer.go:207]: Please add `defer ticker.Stop()` with a note that it can be removed once we move to Go 1.23

File: sandbox/cgroup/freezer.go
Comment (zyga) [inline sandbox/cgroup/freezer.go:103]: Please add `defer ticker.Stop()` with a note to perhaps remove this once we can depend on Go 1.23. See https://pkg.go.dev/time#NewTicker for rationale.

File: sandbox/cgroup/kill.go
Comment (zyga) [reply sandbox/cgroup/kill.go:44]: Perhaps it should be initialized to a dummy function that panics or to nil. Not sure if this is worth it but it might help with casual readability that it's not the final function.

File: sandbox/cgroup/kill.go
Comment (zyga) [inline sandbox/cgroup/kill.go:46]: I think it might be cleaner to just have one toggle v1, v2, fake and have it flip all the functions related to cgroups, so that it's not possible to skew.

**Review by zyga — COMMENTED** [review]
It looks mostly good. Ping me for a final pass.

Separately for landing: I would pehraps consider splitting this into:
- new cgroup code and minimal spill to other packages
- switch over in backend
- removal of userd code
- removal of client code

Those four could be either four patches or four PRs that have messy history and are to be squash merged with commit message rewrite.

File: cmd/snap-update-ns/common_test.go
Comment (ZeyadYasser) [reply cmd/snap-update-ns/common_test.go:27]: Thanks for catching this!

File: cmd/snap-update-ns/common.go
Comment (ZeyadYasser) [reply cmd/snap-update-ns/common.go:83]: Good point, Added a TODO

File: overlord/snapstate/backend/link.go
Comment (ZeyadYasser) [reply overlord/snapstate/backend/link.go:421]: Totally, Added a TODO

File: overlord/snapstate/handlers.go
Comment (ZeyadYasser) [reply overlord/snapstate/handlers.go:3317]: yes, I noticed in `stopSnapServices` and `undoStartSnapServices`, that `backend.StopServices` gets unlocked so I followed the pattern there. I can dig deeper into figuring out if the state unlock is ever needed.

File: sandbox/cgroup/freezer.go
Comment (ZeyadYasser) [reply sandbox/cgroup/freezer.go:207]: Nice catch, Thanks!

File: sandbox/cgroup/freezer.go
Comment (ZeyadYasser) [reply sandbox/cgroup/freezer.go:103]: Good point, it would make tests a bit faster

File: tests/main/snap-remove-terminate/api-client/meta/snap.yaml
Comment (zyga) [inline tests/main/snap-remove-terminate/api-client/meta/snap.yaml:3]: Nitpick, is core18 relevant here? Could we use something less more recent?

**Review by zyga — APPROVED** [review]
LGTM! Nice work :-)

I left one comment about the test snap and the base used there. Please decide if this is something you want to act upon.

File: tests/main/snap-remove-terminate/api-client/meta/snap.yaml
Comment (ZeyadYasser) [reply tests/main/snap-remove-terminate/api-client/meta/snap.yaml:3]: Good point, I have created ticket (SNAPDENG-26852) to merge the several duplicates of `api-client`, I added a comment about addressing the base there as well.

File: sandbox/cgroup/kill.go
Comment (pedronis) [inline sandbox/cgroup/kill.go:90]: nitpick: s/action/killCgroupProcs/

**Review by pedronis — COMMENTED** [review]
thanks, one nitpick and one question

File: overlord/snapstate/handlers.go
Comment (ZeyadYasser) [reply overlord/snapstate/handlers.go:3308]: Good point, no we don't need the state lock. Updated and added a comment.

**Review by pedronis — APPROVED** [review]
thank you

**Comment by ZeyadYasser** [comment]
I am noticing failed freeze attempts in `ubuntu-secboot`, need to debug this and find out if it is just the [known v1 freezer kernel bug](https://github.com/opencontainers/runc/blob/1aeefd9cbdda983d75bdd8d869fe2ac5faab3707/libcontainer/cgroups/fs/freezer.go#L38-L59) or something else implementation related.

Please don't merge this PR.

**Comment by ZeyadYasser** [comment]
I found out the issue, it was a bug that was basically there for 7 years
https://github.com/canonical/snapd/blob/bef2969981588c6ac01881415401b6c852fb65b2/sandbox/cgroup/freezer.go#L95-L106

The kernel returns the `freezer.state` file with a newline suffix, and the check was not trimming whitespace after reading `freezer.state` file. When I switched the implementation in this PR to check to be more aggressive where it checks for the desired `FROZEN` state instead of the `FREEZING` state, the bug surfaced. Previously it wasn't waiting for `FREEZING` state to go away, because it was always getting `FREEZING\n` not the expected `FREEZING`.

**Comment by ZeyadYasser** [comment]
Waiting for test run before merging

**Comment by ZeyadYasser** [comment]
Requesting to force merge, only failing test is `google:ubuntu-20.04-64:tests/main/nvidia-files:535_server` which is known to fail.

**Pull request merged**
